### ISSUE TITLE
求教为什么跑出来的异常图结果不太对

### ISSUE THREAD

**Issue by Lhy-opt** [ZHE-SAPI/CostFilter-AD#13]
您好，向您请教一个问题，在Dinomaly ViTill模型跑异常图的推理结果，代码如下：

...
    model = ViTill(encoder=encoder, bottleneck=bottleneck, decoder=decoder, target_layers=target_layers,
                    mask_neighbor_size=0, fuse_layer_encoder=fuse_layer_encoder, fuse_layer_decoder=fuse_layer_decoder)

    img_test = cv2.imread('000.png')
    img_test = cv2.resize(img_test, (392, 392))
    img_test = cv2.cvtColor(img_test, cv2.COLOR_BGR2RGB)
    img_test = torch.from_numpy(img_test).permute(2, 0, 1).unsqueeze(0).float() / 255.0
    img_test = img_test.to(device)
    with torch.no_grad():
        enc, dec = model(img_test)
        anomaly_map, _ = cal_anomaly_map(enc, dec, img_test.shape[-1], amap_mode='add')
        anomaly_map = np.expand_dims(anomaly_map, -1)
        
        min_val = np.min(anomaly_map)
        max_val = np.max(anomaly_map)
        anomaly_map = (anomaly_map - min_val) / (max_val - min_val)
        anomaly_map = cv2.applyColorMap(np.uint8(anomaly_map * 255), cv2.COLORMAP_JET)
        cv2.imwrite("anomaly_map.png", anomaly_map)

得到的结果如下：
原图
<img width="392" height="392" alt="Image" src="https://github.com/user-attachments/assets/83c0c9a7-1aa8-486d-a69a-2afa2ae29b65" />

异常图

<img width="392" height="392" alt="Image" src="https://github.com/user-attachments/assets/b679acb9-ff38-48a9-ae0b-38449b545c80" />

和论文里说的不太一样，而且我没有对原图做什么特殊处理，拿到编码器和解码器的结果之后只是做了dim1 的余弦相似性，您有什么头绪吗，我不知道这样的异常图是否正确

**Comment by Lhy-opt** [comment]
源代码里的cal_anomaly_map函数定义如下：

def cal_anomaly_map(fs_list, ft_list, out_size=224, amap_mode='add', norm_factor=None):
    if not isinstance(out_size, tuple):
        out_size = (out_size, out_size)
    if amap_mode == 'mul':
        anomaly_map = np.ones(out_size)
    else:
        anomaly_map = np.zeros(out_size)

    a_map_list = []
    for i in range(len(ft_list)):
        fs = fs_list[i]
        ft = ft_list[i]
        a_map = 1 - F.cosine_similarity(fs, ft)
        a_map = torch.unsqueeze(a_map, dim=1)
        a_map = F.interpolate(a_map, size=out_size, mode='bilinear', align_corners=True)
        if norm_factor is not None:
            a_map = 0.1 * (a_map - norm_factor[0][i]) / (norm_factor[1][i] - norm_factor[0][i])

        a_map = a_map[0, 0, :, :].to('cpu').detach().numpy()
        a_map_list.append(a_map)
        if amap_mode == 'mul':
            anomaly_map *= a_map
        else:
            anomaly_map += a_map
    return anomaly_map, a_map_list

**Comment by Lhy-opt** [comment]
感谢您的代码，已成功拿到热力图！

**Comment by ZHE-SAPI** [comment]
不客气，感谢关注！

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2025年08月25日 04:13 |
| 收件人 | ZHE-SAPI/CostFilter-AD ***@***.***> |
| 抄送人 | Zhang ***@***.***>,
Comment ***@***.***> |
| 主题 | Re: [ZHE-SAPI/CostFilter-AD] 求教为什么跑出来的异常图结果不太对 (Issue #13) |
Lhy-opt left a comment (ZHE-SAPI/CostFilter-AD#13)

感谢您的代码，已成功拿到热力图！
